### ISSUE TITLE
Evaluation, list and collaboration with existing blockchains

### ISSUE THREAD

**Issue by juanfranblanco** [hyperledger/hyperledger#7]
Following the discussion on slack.

Before we commit to any specific code base and or contribution, it will be ideal to list the pros and cons of existing blockchains. This will allow to not reinvent the wheel and in cases the square wheel. 

Examples of this will be but not limited to (Please comment or edit to add more blockchains, whom add value based on their new features / implementation)
- Ethereum
- Bitshares / Graphene
- Engima
- MaidSafe
- Dash
- Eris (as a platform)
- Tendermint (as a deterministic building block)

Already included / contributed
- OBC (IBM)
- Ripple
- Digital Assets
- BlockStream

This should included DFS (blockchain based or not) for decentralised storage of both front and backend.
- IPFS
- Tahoe-LAFS
- Storj

SmartContracts:
- CommonAccord

The aim of this is to not commit to any implementation before we have evaluated current solutions (already open source), for performance, scalability, privacy, consensus implementation and development friendly (smart contracts, api, etc). 

Collaboration should be sought from current blockchains and experts if they are not already collaborating.

In plain terms as per slack comment: "let's figure out what hyperledger NEEDs and what is already out there"

**Comment by dazraf** [comment]
Emphatic +1

**Comment by jeffscottward** [comment]
IPFS is not a blockchain, but worth investigating as a storage layer yes

**Comment by juanfranblanco** [comment]
"IPFS is not a blockchain, but worth investigating as a storage layer yes", correct it was added for the decentralization storage capabilities, and it also seems to be becoming the preferred option at the moment.

**Comment by bryceweiner** [comment]
What are the benefits of IPFS over other DFS implementations such as Tahoe-LAFS?

**Comment by juanfranblanco** [comment]
Not biased, added to the list :)

**Comment by dazraf** [comment]
How about:
* Eris
* Tendermint (as a deterministic building block)

**Comment by mdangear** [comment]
If you start getting into building blocks you should also consider
CommonAccord. Smart contracts are not contracts, and CommonAccord allows to
tie actual contracts to transactions, to build the legal context.
Open source, on Github. The data model allows to build document (text) that
can be read both by humans and by machines.

**Comment by dazraf** [comment]
+1 @mdangear

**Comment by juanfranblanco** [comment]
Done, (Please continue mentioning anyone that it might be interesting) I think this needs further pushing on slack channel, as there is no participation from anybody representing Hyperledger only community.

**Comment by greggigon** [comment]
IPFS, interesting addition. ++ from me.
It would be great to have consistency algorithms working over distributed data, making it easier to solve scale problems.

**Comment by jaekwon** [comment]
Hi all, Jae from Tendermint here.  I'd like to suggest a proposal of procedure.

There's a lot of design choices to be made for a blockchain platform, and we at Tendermint have made several decisions already.  For example, the BFT consensus algorithm that we developed, in our opinion, makes certain tradeoffs that are very reasonable, and should suffice for 99% of use-cases, and provides benefits above and beyond what is provided with say, PBFT.

But that's not what I'm here to propose.  I'm here to propose the following:  Instead of deciding on the overall architecture of Hyperledger in one go, lets decide on a single thing first -- whether or not to adopt the Tendermint-TMSP/OBP-Chaincode approach of separating the platform from the business logic via a network socket protocol, and if so, lets decide on what that protocol should be.

The benefit of this approach is that (A) it works with any blockchain application business logic e.g. whether based on UTXO or EVM or whatever, and (B) once we decide on this protocol, we can then work on standardizing the details of the platform and the details of the application in tandem.  In short, it lets us divide the task at hand, and increases the changes that this collaborative project will create real value.

**Comment by rickseeger** [comment]
There are definitely some features in existing blockchains to examine / replicate in Hyperledger:

BitShares - the original business-friendly blockchain, C++ (i.e.fast) at 100,000 tx/sec; supports: bonds, decentralized exchange, price stability through CFDs, and has a robust governance model (so important!)

Ethereum - Turing complete and generic, it's tempting just to fork and rebrand that entire repo and start building on it (e.g. see SHFT, the first fork of Ethereum: https://bitcointalk.org/index.php?topic=1155284.0) Downside: the transactions per second is fairly limited due to block propagation speed.

DASH - very fast tx times, good governance model, but not really built for smart contracts, basically digital cash like bitcoin

MaidSafe/StorJ - cloud storage DAOs, not very relevant here

Emercoin - lots of nice features: DNS, advertising network, SSH, proof of existence, file sharing, ... and it's already proven to function as backbone for blockchain as a service on Azure

**Comment by shemnon** [comment]
There sure are a lot of options out there.  But if Hyperledger wants to be an industry standard like Linux it needs to be built with some customizability in at least scale from the beginning.  Look at Linux (the namesake of the foundation this is being run under).  The demands of desktop, mobile phone, raspberry pi, and cloud servers are quite varied, yet at the core of each is the Linux kernel.  Although in some ways it looks very different in each configuration.

So if there were some level of plugability in core areas that would go a long way to making HL the standard kernel of the blockchain world.

Right now I see plugability needs in three different areas:

Consensus: this is the biggest one with the strongest opinions.  In the IBM contribution there is already plug ability for consensus: PBFT, SEIVE, and none.  Then there are the other industry options of POW and POS.

Payload:  UTXO, EVM, and Chaincode have already been discussed.  But those might be overkill for a simple Proof of Existence system, especially on where the economic motive to maintain it is not on the blockchain itself. But in a block chain every payload needs to be reduced to a hash at some point, so this may be one of the easier things to stub out.

Networking/Distribution: I think this is the stealth issue, but only because the first two are dominating the discussion already.  Not everything benefits from a gossip network, but it is a favorite.  Some models like Storj will need a special emphasis on how to share the payload, and many Proof of Existence models would actively want registered content to be configurable opaque or transparent, and then sometimes transparent only to limited parties (although out of band may be better for limited distribution).

And then to complicate some systems require something out of each of the aspects.  Proof of Stake needs to know the content of the payload and may be running it’s round robin off of the network topology.  So let’s add high cohesion and low coupling to the buzzword stew while we are at it.

But this may be hard to do and have a viable result in a collaborative setup.  Making something good and usable that has plug ability in mind but doesn’t actually have it than something with all the plugs in the world that no one uses.  Just look at your laptop for proof. 

tl;dr Plugability is good.  But it may be a 1.1 or 2.0 target.

**Comment by marcantoineTrehin** [comment]
OneName : Identity Manager

**Comment by shea256** [comment]
Hey guys, very interesting conversation you have going here. Really excited for the development of the hyperledger project.

@marcantoineTrehin thanks for looping me in. You may know me as a co-founder of Onename, which is an application built on Blockstack - a technology that we've been pioneering along with others.

Thought I'd give you some background on Blockstack and you can decide if it's interesting to you.

In a nutshell, Blockstack is a system for decentralized DNS and identity, built as a virtual layer on top of blockchains (currently there's one Blockstack network that operates on the Bitcoin blockchain).

Blockstack has a simple key-value store interface, where globally-unique, human-meaningful names or keys can be registered and owned by a keypair. We're going to be adding more advanced data setting and query operations so that it behaves a bit more like a database.

Blockstack is very versatile and can be used as a replacement for DNS, as an identity system, or as a decentralized replacement for registries and directories.

Here's the documentation on Blockstack: https://blockstack.org/docs, and here's a listing of the namespaces that are in the pipeline: https://blockstack.org/docs/namespaces.

cc'ing @muneeb-ali, @jcnelson, @guylepage3, and @larrysalibra, who also work on Blockstack.

**Comment by shea256** [comment]
I'll also point out the Blockstack is the largest non-financial protocol on the Bitcoin blockchain by a wide margin, as you can see here: http://opreturn.org/

![screen shot 2016-04-01 at 2 05 30 pm](https://cloud.githubusercontent.com/assets/994646/14215744/d98601d4-f812-11e5-9d28-8d11da795ff1.png)

**Comment by shea256** [comment]
@juanfranblanco I'd love to continue this discussion here or in Slack. Can you send me an invite?

**Comment by juanfranblanco** [comment]
@shea256  a bit late on the response.. the slack is open to anyone.

**Comment by Danajoyluck** [comment]
@juanfranblanco is there a new link to sign up for slack? I went to https://slack.hyperledger.org/, it still asks for invite. Many thanks!

**Comment by juanfranblanco** [comment]
Just put your email address and you will be redirected if already signed. Or you will get an email if not.

**Comment by juanfranblanco** [comment]
Here is the presentation about the possibilities of integrating the EVM (Ethereum) into Hyperledger https://drive.google.com/file/d/0B42vMkapQi1MREpMMlUwR2V1UDQ/view

**Comment by yuanlaishiluda** [comment]
IBM zSystem ?

**Comment by exside** [comment]
Maybe add SiaCoin to the storage part too, looks actually better than Storj (they are centralized) and don't forget to have an eye on Ardor (NXT 2.0)

**Comment by chainhead** [comment]
Stellar? Or, it's predecessor Ripple?
